### ISSUE TITLE
"Slice end index out of range" on eval data

### ISSUE THREAD

**Issue by korneil** [dmlc/mxnet#612]
During the eval phase I get this error message:

		[00:43:49] ./dmlc-core/include/dmlc/logging.h:208: [00:43:49] include/mxnet/ndarray.h:219: Check failed: (shape_[0]) >= (end) Slice end index out of range
		Traceback (most recent call last):
		  File "mxnet_xx.py", line 142, in <module>
		    batch_end_callback=mx.callback.Speedometer(100))
		  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/mxnet-0.5.0-py2.7.egg/mxnet/model.py", line 756, in fit
		    logger=logger, work_load_list=work_load_list)
		  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/mxnet-0.5.0-py2.7.egg/mxnet/model.py", line 373, in _train_multi_device
		    _load_label(eval_batch, label_arrays)
		  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/mxnet-0.5.0-py2.7.egg/mxnet/model.py", line 55, in _load_label
		    _load_general(batch.label, targets)
		  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/mxnet-0.5.0-py2.7.egg/mxnet/model.py", line 49, in _load_general
		    d_src[slice_idx].copyto(d_dst)
		  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/mxnet-0.5.0-py2.7.egg/mxnet/ndarray.py", line 219, in __getitem__
		    return self._slice(in_slice.start, in_slice.stop)
		  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/mxnet-0.5.0-py2.7.egg/mxnet/ndarray.py", line 259, in _slice
		    self.handle, start, stop, ctypes.byref(handle)))
		  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/mxnet-0.5.0-py2.7.egg/mxnet/base.py", line 73, in check_call
		    raise MXNetError(py_str(_LIB.MXGetLastError()))
		mxnet.base.MXNetError: [00:43:49] include/mxnet/ndarray.h:219: Check failed: (shape_[0]) >= (end) Slice end index out of range

Before ndarray.h:~219:

        LOG(INFO) << "slice.begin "<<begin<<" slice.end "<<end<<" shape_[0]: "<< shape_[0];

        [00:48:35] include/mxnet/ndarray.h:217: slice.begin 0 slice.end 16 shape_[0]: 15

If I fix the shape in ndarray.h, later it complains about shape mismatch.

The problem is around model.py:~250:

        label_arrays = [[(slices[i], e.arg_dict[name]) for i, e in enumerate(train_execs)]
                    for name in label_names]

or more likely with the `slices` generation itself in `_split_input_slice()`

**Comment by korneil** [comment]
Solved, it was an error on my side, calling `mx.io.NDArrayIter` with different shapes.

**Comment by diggerdu** [comment]
what are you mean?
please explain it more in detail
i was screwed by this probleam

**Comment by thvasilo** [comment]
For me the case was that I had trained the model using one batch size and then tried to evaluate using a different one.

E.g.

```python
train_iter = mx.io.NDArrayIter(X_train, y_train, batch_size, shuffle=True)
test_iter = mx.io.NDArrayIter(X_test, y_test, batch_size)
```

Note how I assign the same batch size to both iterators. When I tried using different batch sizes it would error out. Not sure if this is a bug or not, I don't see why the batch size needs to be tied to the model after training.
